File: app-android/app/src/main/java/de/tutao/tutanota/credentials/CredentialsEncryptionFactory.java
package de.tutao.tutanota.credentials;

import android.os.Build;

import de.tutao.tutanota.AndroidKeyStoreFacade;
import de.tutao.tutanota.AndroidKeyStoreFacadeFactory;
import de.tutao.tutanota.MainActivity;

/**
 * We use a factory for this to be able to cleanly separate code for different Android API versions.
 */
public final class CredentialsEncryptionFactory {
	public static ICredentialsEncryption create(MainActivity activity) {
		AuthenticationPrompt authenticationPrompt = new AuthenticationPrompt();
		AndroidKeyStoreFacade keyStoreFacade = AndroidKeyStoreFacadeFactory.create(activity);
		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
			return new CredentialsEncryptionBeforeAPI30(keyStoreFacade, activity, authenticationPrompt);
		} else {
			return new CredentialsEncryptionFromAPI30(keyStoreFacade, activity, authenticationPrompt);
		}
	}
}

File: app-android/app/src/main/java/de/tutao/tutanota/credentials/CredentialsEncryptionFactory.kt
package de.tutao.tutanota.credentials

import android.content.Context
import android.os.Build
import de.tutao.tutanota.AndroidNativeCryptoFacade
import de.tutao.tutanota.data.AppDatabase
import de.tutao.tutanota.ipc.NativeCredentialsFacade

/**
 * We use a factory for this to be able to cleanly separate code for different Android API versions.
 */
object CredentialsEncryptionFactory {
	fun create(activity: Context, crypto: AndroidNativeCryptoFacade, db: AppDatabase): NativeCredentialsFacade {
		val authenticationPrompt = AuthenticationPrompt()
		val keyStoreFacade = de.tutao.tutanota.createAndroidKeyStoreFacade()
		val keychainEncryption = if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
			CredentialsEncryptionBeforeAPI30(keyStoreFacade, activity, crypto, authenticationPrompt)
		} else {
			CredentialsEncryptionFromAPI30(keyStoreFacade, activity, crypto, authenticationPrompt)
		}
		return AndroidNativeCredentialsFacade(crypto, keychainEncryption, db)
	}
}